src/main/java/io/vertx/ext/stomp/impl/DefaultSendHandler.java

```diff
@@ -0,0 +1,112 @@
+ * sent to all matching subscriptions.
+ * <p/>
+ * If the {@code SEND} frame requires an acknowledgment, the {@code message-id} is added to the list of messages
+ * waiting for acknowledgment. If the corresponding {@code ACK} frame does not arrives before
+ * {@link StompServerOptions#getAckTimeout()} ms, the message is considered as non-acknowledged.
+ *
+ * This handler is thread safe.
+ */
+public class DefaultSendHandler implements ServerFrameHandler {
+
+  private static final Logger log = LoggerFactory.getLogger(DefaultSendHandler.class);
+
+
+  @Override
+  public void onFrame(Frame frame, StompServerConnection connection) {
+    String destination = frame.getHeader(Frame.DESTINATION);
+    if (destination == null) {
+      connection.write(Frames.createErrorFrame(
+          "Destination header missing",
+          Headers.create(frame.getHeaders()), "Invalid send frame - the " +
+              "'destination' must be set"));
+      connection.close();
+      return;
+    }
+
+    // Handle transaction
+    String txId = frame.getHeader(Frame.TRANSACTION);
+    if (txId != null) {
+      Transaction transaction = connection.handler().getTransaction(connection, txId);
+      if (transaction == null) {
+        // No transaction.
+        Frame errorFrame = Frames.createErrorFrame(
+            "No transaction",
+            Headers.create(Frame.DESTINATION, destination, Frame.TRANSACTION, txId),
+            "Message delivery failed - unknown transaction id");
+        connection.write(errorFrame);
+        connection.close();
+        return;
+      } else {
+        transaction.addFrameToTransaction(frame);
+        Frames.handleReceipt(frame, connection);
+        // No delivery in transactions.
+        return;
+      }
+    }
+
+
+    List<Subscription> subscriptions = connection.handler().getSubscriptions(destination);
+    if (subscriptions.isEmpty() && connection.server().getOptions().isSendErrorOnNoSubscriptions()) {
+      Frame errorFrame = Frames.createErrorFrame(
+          "No subscriptions",
+          Headers.create(Frame.DESTINATION, destination),
+          "Message delivery failed - no subscriptions on this destination");
+      connection.write(errorFrame);
+      connection.close();
+      return;
+    }
+
+    subscriptions.stream().forEach(subscription -> {
+          String messageId = UUID.randomUUID().toString();
+          Frame message = sendToMessage(frame, subscription, messageId);
+          enqueue(connection, subscription, message);
+          subscription.connection().write(message.toBuffer());
+        }
+    );
+
+    Frames.handleReceipt(frame, connection);
+  }
+
+  private void enqueue(StompServerConnection connection, Subscription subscription, Frame frame) {
+    subscription.enqueue(frame);
+    if (connection.server().getOptions().getAckTimeout() != 0) {
+      long time = connection.server().getOptions().getAckTimeout() * connection.server().getOptions().getTimeFactor();
```

Comment:
Setting a new timer for each message is not very efficient. Would be better to have a single timer (or maybe one per subscription) that scans the messages in the subscriptions periodically. Although I'm not sure we need a timeout at all (I never saw this in the messaging systems I wrote/worked with before).